python/cudf/dataframe/dataframe.py

```diff
@@ -855,6 +855,90 @@ def index(self, _index):
         for k in self.columns:
             self[k] = self[k].set_index(idx)
 
+    def reindex(self, labels=None, axis=0, index=None, columns=None,
+                copy=True):
+        """Return a new DataFrame whose axes conform to a new index
+
+        ``DataFrame.reindex`` supports two calling conventions
+        * ``(index=index_labels, columns=column_names)``
+        * ``(labels, axis={0 or 'index', 1 or 'columns'})``
+
+        Parameters
+        ----------
+        labels : Index, Series-convertible, optional, default None
+        axis : {0 or 'index', 1 or 'columns'}, optional, default 0
+        index : Index, Series-convertible, optional, default None
+            Shorthand for ``df.reindex(labels=index_labels, axis=0)``
+        columns : array-like, optional, default None
+            Shorthand for ``df.reindex(labels=column_names, axis=1)``
+        copy : boolean, optional, default True
+
+        Returns
+        -------
+        A DataFrame whose axes conform to the new index(es)
+
+        Examples
+        --------
+        >>> import cudf
+        >>> df = cudf.DataFrame()
+        >>> df['key'] = [0, 1, 2, 3, 4]
+        >>> df['val'] = [float(i + 10) for i in range(5)]
+        >>> df_new = df.reindex(index=[0, 3, 4, 5],
+                                columns=['key', 'val', 'sum'])
+        >>> print(df)
+           key   val
+        0    0  10.0
+        1    1  11.0
+        2    2  12.0
+        3    3  13.0
+        4    4  14.0
+        >>> print(df_new)
+           key   val  sum
+        0    0  10.0  NaN
+        3    3  13.0  NaN
+        4    4  14.0  NaN
+        5   -1   NaN  NaN
+        """
+
+        if labels is None and index is None and columns is None:
+            return self.copy(deep=copy)
+
+        df = self
+        cols = columns
+        dtypes = OrderedDict(df.dtypes)
+        idx = labels if index is None and axis in (0, 'index') else index
+        cols = labels if cols is None and axis in (1, 'columns') else cols
+        df = df if cols is None else df[list(set(df.columns) & set(cols))]
+
+        if idx is not None:
+            idx = idx if isinstance(idx, Index) else as_index(idx)
+            if df.index.dtype != idx.dtype:
+                df = DataFrame()
+            else:
+                df = DataFrame(None, idx).join(df, how='left', sort=True)
+                # TODO: use df.take() once string cols are working
+                # df = df.take(idx.argsort(True).argsort(True).to_gpu_array())
```

Comment:
@kkraus14 we should use `df.take()` here. `cpp_copy.apply_gather()` uses `columnops.column_empty()` to create the out columns, but 1. it doesn't update the `null_count` afterwards, and 2. doesn't seem to write the nvstrings nullmask.

Reply:
> @kkraus14 we should use `df.take()` here, but I ran into two issues: `cpp_copy.apply_gather()` uses `columnops.column_empty()` to create the out columns, but it doesn't update the `null_count` afterwards, and doesn't seem to write the nvstrings nullmask.

It looks like the StringColumn initialization function should handle calculating the null count and updating the nullmask. Are you talking about using Strings in the index?

Reply:
No, I mean I think I was running into a bug in the cpp apply_gather itself. I'd need to look back at my notes, but here's roughly what I was seeing:

```python
df = DataFrame({'str': ['foo', 'bar', 'baz']})
df2 = df.take([0, 2])
print(df2['str'].null_count)
# 2

print(df2.to_pandas())
#   str
# 0 None
# 1 None
```

Reply:
@shwina did you run into this issue and fix the gather code for NVCategory objects?

Reply:
@kkraus14 ok, here's a repro of the issue I encountered:

<details><summary>Click to expand python code</summary><pre lang="python">
import cudf
import asyncio
#
print(cudf.__version__)
print(cudf)
#
df = cudf.datasets.randomdata(nrows=6, dtypes={'a': 'category', 'c': float, 'd': str})
#
print('test dataset:\n', df.to_pandas())
#
index = cudf.dataframe.index.as_index([-1, 0, 3])
df2 = cudf.DataFrame(None, index).join(df, how='left', sort=True)
positions = cudf.Series(index.argsort(True).argsort(True).to_gpu_array())
#
print('join result:\n', df2.to_pandas())
print('original indices:\n', positions.to_pandas())
print('reindexed result:\n', df2.take(positions).to_pandas())
#
# the error happens asynchronously the first time
await asyncio.sleep(1)
#
df3 = cudf.datasets.randomdata(nrows=6, dtypes={'a': 'category', 'c': float, 'd': str})
#
print('this blows up:\n', df3.to_pandas())
</pre></details>

<details><summary>Click to expand output</summary><pre lang="text/plain">
0+unknown
module 'cudf' from '/opt/rapids/cudf/python/cudf/__init__.py'
test dataset:
         a         c       d
0  Hannah -0.896236     Ray
1     Dan  0.803338   Quinn
2     Tim -0.720630   Laura
3   Zelda -0.593139   Frank
4   Sarah  0.754045  Ursula
5     Ray -0.804741  Hannah
join result:
          a         c      d
-1     NaN       NaN   None
 0  Hannah -0.896236    Ray
 3   Zelda -0.593139  Frank
original indices:
 0    0
1    1
2    2
dtype: int32
reindexed result:
          a         c     d
-1     NaN       NaN  None
 0  Hannah -0.896236  None
 3   Zelda -0.593139  None
---------------------------------------------------------------------------
</pre></details>

<details><summary>Click to expand error stack trace</summary><pre>
RuntimeError                              Traceback (most recent call last)
<\ipython-input-4-d588ebb20c1f> in async-def-wrapper()
     22 print('this blows up:\n', df3.to_pandas())

/opt/rapids/cudf/python/cudf/datasets.py in randomdata(nrows, dtypes, seed)
     84     columns = dict((k, make[dt](nrows, state)) for k, dt in dtypes.items())
     85     df = pd.DataFrame(columns, columns=sorted(columns))
---> 86     return cudf.from_pandas(df)
     87 
     88 

/opt/rapids/cudf/python/cudf/dataframe/dataframe.py in from_pandas(obj)
   3166     """
   3167     if isinstance(obj, pd.DataFrame):
-> 3168         return DataFrame.from_pandas(obj)
   3169     elif isinstance(obj, pd.Series):
   3170         return Series.from_pandas(obj)

/opt/rapids/cudf/python/cudf/dataframe/dataframe.py in from_pandas(cls, dataframe, nan_as_null)
   2629             # columns for a single key
   2630             if len(vals.shape) == 1:
-> 2631                 df[colk] = Series(vals, nan_as_null=nan_as_null)
   2632             else:
   2633                 vals = vals.T

/opt/rapids/cudf/python/cudf/dataframe/series.py in __init__(self, data, index, name, nan_as_null, dtype)
     86         if not isinstance(data, columnops.TypedColumnBase):
     87             data = columnops.as_column(data, nan_as_null=nan_as_null,
---> 88                                        dtype=dtype)
     89 
     90         if index is not None and not isinstance(index, Index):

/opt/rapids/cudf/python/cudf/dataframe/columnops.py in as_column(arbitrary, nan_as_null, dtype)
    309             data = as_column(pa.Array.from_pandas(arbitrary))
    310         else:
--> 311             data = as_column(rmm.to_device(arbitrary), nan_as_null=nan_as_null)
    312 
    313     elif isinstance(arbitrary, pa.Array):

/opt/rapids/cudf/python/cudf/dataframe/columnops.py in as_column(arbitrary, nan_as_null, dtype)
    286             if nan_as_null:
    287                 mask = cudautils.mask_from_devary(arbitrary)
--> 288                 data = data.set_mask(mask)
    289 
    290     elif cuda.is_cuda_array(arbitrary):

/opt/rapids/cudf/python/cudf/dataframe/column.py in set_mask(self, mask, null_count)
    263             msg = 'mask must be of byte; but got {}'.format(mask.dtype)
    264             raise ValueError(msg)
--> 265         return self.replace(mask=mask, null_count=null_count)
    266 
    267     def allocate_mask(self, all_valid=True):

/opt/rapids/cudf/python/cudf/dataframe/numerical.py in replace(self, **kwargs)
     48         if 'data' in kwargs and 'dtype' not in kwargs:
     49             kwargs['dtype'] = kwargs['data'].dtype
---> 50         return super(NumericalColumn, self).replace(**kwargs)
     51 
     52     def serialize(self, serialize):

/opt/rapids/cudf/python/cudf/dataframe/column.py in replace(self, **kwargs)
    376         if 'mask' in kwargs and 'null_count' not in kwargs:
    377             del params['null_count']
--> 378         return type(self)(**params)
    379 
    380     def view(self, newcls, **kwargs):

/opt/rapids/cudf/python/cudf/dataframe/numerical.py in __init__(self, **kwargs)
     42             Data type
     43         """
---> 44         super(NumericalColumn, self).__init__(**kwargs)
     45         assert self._dtype == self._data.dtype
     46 

/opt/rapids/cudf/python/cudf/dataframe/columnops.py in __init__(self, **kwargs)
     38     def __init__(self, **kwargs):
     39         dtype = kwargs.pop('dtype')
---> 40         super(TypedColumnBase, self).__init__(**kwargs)
     41         # Logical dtype
     42         self._dtype = pd.api.types.pandas_dtype(dtype)

/opt/rapids/cudf/python/cudf/dataframe/column.py in __init__(self, data, mask, null_count)
    158             assert mask.size * utils.mask_bitsize >= len(self)
    159 
--> 160         self._update_null_count(null_count)
    161 
    162     def _update_null_count(self, null_count=None):

/opt/rapids/cudf/python/cudf/dataframe/column.py in _update_null_count(self, null_count)
    166                 nnz = count_nonzero_mask(
    167                     self._mask.mem,
--> 168                     size=len(self)
    169                 )
    170                 null_count = len(self) - nnz

/opt/rapids/cudf/python/cudf/bindings/cudf_cpp.pyx in cudf.bindings.cudf_cpp.count_nonzero_mask()

/opt/rapids/cudf/python/cudf/bindings/cudf_cpp.pyx in cudf.bindings.cudf_cpp.count_nonzero_mask()

RuntimeError: CUDA error encountered at: /opt/rapids/cudf/cpp/src/bitmask/bitmask_ops.cu:147: 17 cudaErrorInvalidDevicePointer invalid device pointer
</pre></details>